File: competitive-programming/src/edu/algos/li13_priorityqueue/_L703_KthLargestElementInAStream.java
package edu.algos.li13_priorityqueue;

import java.util.PriorityQueue;

public class _L703_KthLargestElementInAStream {

	public static void main(String[] args) {
		_L703_KthLargestElementInAStream kthLargest = new _L703_KthLargestElementInAStream(3, new int[] {4, 5, 8, 2});
		System.out.println(kthLargest.add(3));
		System.out.println(kthLargest.add(5));
		System.out.println(kthLargest.add(10));
		System.out.println(kthLargest.add(9));
		System.out.println(kthLargest.add(4));
	}
	
    // Problem: https://leetcode.com/problems/kth-largest-element-in-a-stream/
    // Idea: Keep k largest elements in min heap so that top of the pq is kth largest.
	int k;
    private PriorityQueue<Integer> minHeap = new PriorityQueue<Integer>();
    public _L703_KthLargestElementInAStream(int k, int[] nums) {
        this.k = k;
        for(int num: nums) {
        	minHeap.add(num);
        }
        while(minHeap.size() > k) {
        	minHeap.poll();
        }
    }
    
    public int add(int val) {
        if(minHeap.size() < k){
            minHeap.add(val);
        }else if(val > minHeap.peek()) {
        	minHeap.add(val);
        	minHeap.poll();
        }
        
    	if(minHeap.size() < k){
            return -1;
        }
        return minHeap.peek();
    }
}

File: dsa/patterns/11_heaps/python/l703_kth_largest_element_in_astream.py
import heapq

"""
Problem: https://leetcode.com/problems/kth-largest-element-in-a-stream/
Idea: Keep k largest elements in min heap so that top of the heap is kth largest.
"""
class KthLargest:
    def __init__(self, k, nums):
        self.k = k
        self.min_heap = []

        for num in nums:
            heapq.heappush(self.min_heap, num)
            if len(self.min_heap) > k:
                heapq.heappop(self.min_heap)

    def add(self, val):
        heapq.heappush(self.min_heap, val)
        if len(self.min_heap) > self.k:
            heapq.heappop(self.min_heap)

        if len(self.min_heap) < self.k:
            return -1

        return self.min_heap[0]


# ---- one test (similar to Java main) ----
if __name__ == "__main__":
    kthLargest = KthLargest(3, [4, 5, 8, 2])
    print(kthLargest.add(3))   # 4
    print(kthLargest.add(5))   # 5
    print(kthLargest.add(10))  # 5
    print(kthLargest.add(9))   # 8
    print(kthLargest.add(4))   # 8
